### DISCUSSION TITLE
How to get current Address Wallet?

### DISCUSSION THREAD

**Discussion by nqhhdev** [WalletConnect/WalletConnectFlutterV2#248] [Q&A]
**Describe the bug**
How to get a current Address Wallet?

**Expected behavior**
I want to call the exact wallet currently selected

WalletConnectFlutterV2 version

walletconnect_flutter_v2: 2.1.11
Smartphone (please complete the following information):

Device: Iphone 15 promax
OS: iOS
Dapp Metask v7.12.3 (1230)

**Comment by quetool** [comment]
Hello @nqhhdev ! Could you elaborate a bit more?

**Reply by nqhhdev** [reply]
Hello @quetool 

1. In Metamask I have many wallet addresses (A, B, C, D) and current wallet selected is C
2. I want to send a request with `personal_sign` method using wallet C.
3. I can't know exactly which position is being selected.
4. I expected to know exactly  selected wallet

**Reply by quetool** [reply]
Oh I get the point now. Nope, sadly not, that is something that should happen on wallet side. So in this case, when presented with this screen, Metamask should allow you to select the account you want to sign with.

<img src='https://github.com/WalletConnect/WalletConnectFlutterV2/assets/14978705/b7de929d-d46c-4958-96b4-6f8294b419a1' width='350'>

**Reply by nqhhdev** [reply]
Could you share your code for me?

**Reply by quetool** [reply]
What code?

**Reply by nqhhdev** [reply]
What happens when you send a request with wallet 1 or 2 and in MM wallet selected is 3?

**Reply by nqhhdev** [reply]
```dart
  Future<String?> personalSign() async {
    try {
      final String encodedUri = Uri.encodeComponent('$_uri');

      await launchUrlString(
        'metamask://wc?uri=$encodedUri',
        mode: LaunchMode.externalApplication,
      );

      final result = await _signClient!.request(
        topic: _sessionData!.first.topic,
        chainId: "eip155:1",
        request: SessionRequestParams(
          method: 'personal_sign',
          params: [
            messageSign,
            _sessionData!.first.namespaces["eip155"]?.accounts.first
                .split(":")
                .last
          ],
        ),
      );
      log("personalSign: $result");
      final signature = result as String;
      return signature;
    } catch (e) {
      log(e.toString());
    }
    return null;
  }
```

this is my code. I don't know exactly wallet so I always use wallet first.

**Reply by quetool** [reply]
I don't have a code (your one looks fine by me) but I can tell you that you can not decide, from your side, which account to use. It will always be used the one currently selected on Metamask unless Metamask implement things different. It is not in our hands to be able to SWITCH to a different account since, from a Metamask POV, the selected account is the only active account at that given moment.

The flow is like this:
1. When you connect with MM, your session will be tied to the current active account
2. If you sign a message, THAT active account will be used.
3. If you want to sign with a different account then you should first switch to that account ON METAMASK
4. At that point, Metamask will send an `accountChanged` event to your dapp and now your session will be tied to that new account
5. You now sign a message and that new account will be used.

Just to be clear again, this is not on us (WalletConnect), if you want to be able to switch to a different account during signing you should request this to Metamask https://github.com/MetaMask/metamask-mobile

I hope it's a bit more clear now :)

**Reply by quetool** [reply]
By the way, instead of doing this
```dart
_sessionData!.first.namespaces["eip155"]?.accounts.first.split(":").last
```

You can just use our `NamespaceUtils` class with handy methods:
```dart
final namespace = _sessionData?.namespaces['eip155'];
final accounts = namespace?.accounts ?? [];
final account = NamespaceUtils.getAccount(accounts.first)
```

**Reply by nqhhdev** [reply]
I just sent a request to sign a message with a currently selected wallet on MM.
```dart
_sessionData!.first.namespaces["eip155"]?.accounts.first.split(":").last
```

It's very badly, Is there any other better way?

**Reply by quetool** [reply]
Nope. And again, it's not on our hands.

**Reply by quetool** [reply]
In any case this is a common behavior across every wallet I know.

**Reply by nqhhdev** [reply]
Thanks for your reply. And I have a question for you. I tested it on Android


Case 1:
1. App MM is not open
2. Send a request `personalSign` same my code
3. Don't show a sign popup


Case 2:
1. App MM is open
2. Send a request `personalSign` same my code
3. Show a sign popup

Could you help me find a way better?

**Reply by quetool** [reply]
Is the same happening with any other wallet?

**Reply by nqhhdev** [reply]
Only use MM, sorry 😔

**Reply by quetool** [reply]
Then I can't help you. Every wallet is responsible for handling properly the calls from the dapp using walletconnect protocol.

By the way, I went over your code again and you don't need the encodedUri before signing a message. The only time you need pass the encodedUri to the wallet is when you are pairing with it.

For personalSign you can do just

```dart
Future<String?> personalSign() async {
    try {
      final String encodedUri = Uri.encodeComponent('$_uri');

      launchUrlString(
        'metamask://',
        mode: LaunchMode.externalApplication,
      ); // WITHOUT AWAIT

      final result = await _signClient!.request(
        topic: _sessionData!.first.topic,
        chainId: "eip155:1",
        request: SessionRequestParams(
          method: 'personal_sign',
          params: [
            messageSign,
            _sessionData!.first.namespaces["eip155"]?.accounts.first
                .split(":")
                .last
          ],
        ),
      );
      log("personalSign: $result");
      final signature = result as String;
      return signature;
    } catch (e) {
      log(e.toString());
    }
    return null;
  }
```

You can rely on the redirect object inside `session.peer.metadata`

**Reply by quetool** [reply]
Please remember to mark the answer so we can close the discussion. Thanks!
https://github.com/WalletConnect/WalletConnectFlutterV2/discussions/248#discussioncomment-8108185
